specs/latest/1.0/index.html

```diff
@@ -919,6 +926,50 @@ <h4>Context creation parameters</h4>
                 </p>
               </div>
             </dd>
+        <dt><span class=prop-value>pixelFormat</span></dt>
+            <dd>
+              <p>
+                If defined, require a backbuffer of a specific effective internal format.
+                This is a requirement, and if the requested format is not availible, context creation will fail.
```

Comment:
This would be the first time a WebGL context creation attribute can cause a failure to create a context. Here's why the working group decided previously to not do this, and fall back to supported attributes:

If a context of the same type (webgl, webgl2) has already been created for the canvas / OffscreenCanvas, https://html.spec.whatwg.org/multipage/canvas.html#dom-canvas-getcontext specifies that it has to be returned, regardless of the context creation attributes.

To handle composability of WebGL libraries, where any one could have created the context, this means that any caller of `getContext('webgl[2]', ...)` has to check the context creation attributes to make sure they got what they want.

Having the caller now additionally check for `null` is a little ambiguous - the only way to know that `getContext` returned `null` would be to check for a `webglcontextcreationerror` thrown against the canvas element, which would make context creation extremely contorted and asynchronous.

I appreciate that there's a feature detection problem here, and that some applications might want to fall back in very specific ways if they can't get their desired pixel format.

Do you think there's a different way to handle this? What if we added a setPixelFormat API call at the same time which would be defined as reallocating and zeroing out the back buffer? That one could potentially throw exceptions or similar. Since the context would have been created already, users would have the option of explicitly checking for extension support (for float16 renderability), too.

Reply:
failIfMajorPerformanceCaveat definitely causes failure to create context!

Reply:
With a setPixelFormat-like procedural api, I could see us getting closer to a headless design:
```
const gl = new WebGLRenderContext()
gl.backbufferStorage(gl.RGBA16F, W, H);
```

Reply:
This is #3222

Reply:
Fair point about failIfMajorPerformanceCaveat - had forgotten about that.

Can consider this - but let's discuss on #3222 , since that might end up being a simpler and more useful form of this functionality.

Reply:
> Do you think there's a different way to handle this? What if we added a setPixelFormat API call at the same time which would be defined as reallocating and zeroing out the back buffer? That one could potentially throw exceptions or similar. Since the context would have been created already, users would have the option of explicitly checking for extension support (for float16 renderability), too.

On some underlying implementations (e.g. OpenGL ES) switching to an SRGB_ALPHA8 render buffer will require creating a new context. Are your implementations happy to have to tear down the context and make a new one under the covers in response to setPixelFormat?

Reply:
Well, in Gecko we rarely actually use the default framebuffer for webgl contexts, and when we do it's something we make for ourselves (e.g. EGLSurfaces for an ID3D11Texture). Most of the time we request "headless" contexts when our egl equivalent supports it. I'm not worried about this aspect, but thanks for mentioning it!